payments/providers/talpa_ecom.py

```diff
@@ -0,0 +1,445 @@
+                    "ordinal": "2",
+                }
+            )
+        if length:
+            meta_fields.append(
+                {
+                    "key": "placeLength",
+                    "value": str(length),
+                    "visibleInCheckout": True,
+                    "ordinal": "3",
+                }
+            )
+        if mooring_type:
+            meta_fields.append(
+                {
+                    "key": "placeMooring",
+                    "value": mooring_type,
+                    "visibleInCheckout": True,
+                    "ordinal": "4",
+                }
+            )
+
+        self.payload_add_to_items(payload, order, product_name, meta_fields)
+
+    def payload_add_customer(self, payload: dict, order: Order):
+        if (
+            hasattr(order, "lease")
+            and order.lease is not None
+            and order.lease.application
+        ):
+            application = order.lease.application
+            customer = {
+                "firstName": application.first_name.capitalize(),
+                "lastName": application.last_name.capitalize(),
+                "email": application.email.strip(),
+                "phone": application.phone_number.strip(),
+            }
+        else:
+            customer = {
+                "firstName": order.customer_first_name.capitalize(),
+                "lastName": order.customer_last_name.capitalize(),
+                "email": order.customer_email.strip(),
+                "phone": order.customer_phone.strip(),
+            }
+
+        payload["customer"] = customer
+
+    def handle_settle_order(self, order: Order, event_type: str) -> HttpResponse:
+        status: Optional[OrderStatus]
+        message: str
+        payment_status: str
+
+        if event_type == TALPA_ECOM_WEBHOOK_EVENT_PAYMENT_PAID:
+            status = OrderStatus.PAID
+            message = "Payment succeeded in Talpa eCommerce"
+            payment_status = TalpaEComPaymentDetails.PAYMENT_PAID
+        elif event_type == TALPA_ECOM_WEBHOOK_EVENT_ORDER_CANCELLED:
+            status = OrderStatus.REJECTED
+            message = "Payment rejected in Talpa eCommerce"
+            payment_status = TalpaEComPaymentDetails.PAYMENT_CANCELLED
+        else:
+            raise UnknownWebhookEventError(_("Unknown webhook event: ") + event_type)
+
+        # To verify that the payment information is correct and the call to the webhook
+        # is not being faked, we check that the payment status matches the type of event
+        # that we're getting on the webhook
+        payment_details = self.get_payment_details(order)
+        if payment_details.status != payment_status:
+            return HttpResponseBadRequest()
+
+        try:
+            order.set_status(status, message)
+        except OrderStatusTransitionError as oste:
+            logger.warning(oste)
+            order.create_log_entry(None, comment="Payment succeeded in Talpa eCommerce")
```

Comment:
It might be also cancelled here, not necessarily succeeded?

Reply:
Nice catc 😸